skbio/io/format/gff3.py

```diff
@@ -0,0 +1,534 @@
+        str of seq id, IntervalMetadata
+    '''
+    id_lengths = {}
+    for data_type, sid, data in _yield_record(fh):
+        if data_type == 'length':
+            # get length from sequence-region pragma.
+            # the pragma lines are always before the real annotation lines.
+            id_lengths[sid] = data
+        elif data_type == 'data':
+            length = id_lengths.get(sid)
+            yield sid, _parse_record(data, length)
+
+
+@gff3.writer(None)
+def _generator_to_gff3(obj, fh, skip_subregion=True):
+    '''Write list of IntervalMetadata into file.
+
+    Parameters
+    ----------
+    obj : Iterable of (seq_id, IntervalMetadata)
+    fh : file handler
+    '''
+    # write file header
+    fh.write('##gff-version 3\n')
+    for seq_id, obj_i in obj:
+        _serialize_interval_metadata(obj_i, seq_id, fh, skip_subregion)
+
+
+@gff3.reader(Sequence)
+def _gff3_to_sequence(fh, seq_num=1):
+    ''''''
+    return _construct_seq(fh, Sequence, seq_num)
+
+
+@gff3.writer(Sequence)
+def _sequence_to_gff3(obj, fh, skip_subregion=True):
+    # write file header
+    fh.write('##gff-version 3\n')
+    _serialize_seq(obj, fh, skip_subregion)
+
+
+@gff3.reader(DNA)
+def _gff3_to_dna(fh, seq_num=1):
+    ''''''
+    return _construct_seq(fh, DNA, seq_num)
+
+
+@gff3.writer(DNA)
+def _dna_to_gff3(obj, fh, skip_subregion=True):
+    # write file header
+    fh.write('##gff-version 3\n')
+    _serialize_seq(obj, fh, skip_subregion)
+
+
+@gff3.reader(IntervalMetadata)
+def _gff3_to_interval_metadata(fh, seq_id):
+    '''Read a GFF3 record into the specified interval metadata.
+
+    Parameters
+    ----------
+    fh : file handler
+    '''
+    length = None
+    for data_type, sid, data in _yield_record(fh):
+        if seq_id == sid:
+            if data_type == 'length':
+                # get length from sequence-region pragma
+                length = data
+            elif data_type == 'data':
+                return _parse_record(data, length)
+            else:
+                raise GFF3FormatError(
+                    'Unknown section in the input GFF3 file: '
+                    '%r %r %r' % (data_type, sid, data))
+    # return an empty instead of None
+    return IntervalMetadata(None)
+
+
+@gff3.writer(IntervalMetadata)
+def _interval_metadata_to_gff3(obj, fh, seq_id, skip_subregion=True):
+    '''
+    Parameters
+    ----------
+    obj : IntervalMetadata
+    seq_id : str
+        ID for column 1 in the GFF3 file.
+    '''
+    # write file header
+    fh.write('##gff-version 3\n')
+    _serialize_interval_metadata(obj, seq_id, fh, skip_subregion=True)
+
+
+def _construct_seq(fh, constructor=DNA, seq_num=1):
+    lines = []
+    for i, (data_type, seq_id, l) in enumerate(_yield_record(fh), 1):
+        if data_type == 'data' and seq_num == i:
+            lines = l
+    # you can't read directly from fh because in registry.py line 543
+    # file.tell() will fail "telling position disabled by next() call".
```

Comment:
That sounds right, `.tell()` is disable in python 3 after iteration (which it should be!) because `tell` reflects the C read-ahead buffer which is wholly unrelated to what the iterator of the `for` loop is up to.

Since we are composing filehandles in non-trivial ways, it's not too strange to see this happening. The registry is invoked to make `fh` in the first place, and then invoked again on the `read` call. It is using `tell` and `seek` to try and preserve the `fh` as best it can, this is supposed to be happening before the file-handle is really touched. It is kind of an unfortunate situation as `tell` is only accurate if you seek without reading. Ideally the registry would reset the position to `0` instead of using `tell`, but then this kind of composition would be impossible anyway as it wouldn't ever be able to start in the middle of a file. Meaning you'd still need to split (just like `StringIO` below). Really the only way to avoid this is to not compose sniffing, or invent our own file-handle object in `C` which reflects the Python pointer position instead of the buffer.

Given that, I think we can avoid this if we were to disable format verification, so if you set `verify=False`, you should be able to use `fh` directly because skbio would have no reason to reset the handle's position (because it never sniffed).